src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs

```diff
@@ -56,90 +57,99 @@ internal partial class ServiceConnection : ServiceConnectionBase
-            }
-
-            return Task.CompletedTask;
-        }
-
-        protected override Task OnClientDisconnectedAsync(CloseConnectionMessage closeConnectionMessage)
-        {
-            var connectionId = closeConnectionMessage.ConnectionId;
+        }
+
+        protected override Task<ConnectionContext> CreateConnection(string target = null)
+        {
+            return _connectionFactory.ConnectAsync(HubEndpoint, TransferFormat.Binary, ConnectionId, target, headers: CustomHeader);
+        }
+
+        protected override Task DisposeConnection(ConnectionContext connection)
+        {
+            return _connectionFactory.DisposeAsync(connection);
+        }
+
+        protected override async Task CleanupClientConnections(string fromInstanceId = null)
+        {
+            // To gracefully complete client connections, let the client itself owns the connection lifetime 
+            try
+            {
+                if (_connectionIds.Count == 0)
+                {
+                    return;
+                }
+                var connectionIds = _connectionIds.Select(s => s.Key);
+                if (!string.IsNullOrEmpty(fromInstanceId))
+                {
+                    connectionIds = _connectionIds.Where(s => s.Value == fromInstanceId).Select(s => s.Key);
+                }
+
+                await Task.WhenAll(connectionIds.Select(PerformDisconnectAsyncCore));
+            }
+            catch (Exception ex)
+            {
+                Log.FailedToCleanupConnections(Logger, ex);
+            }
+        }
+
+        protected override ReadOnlyMemory<byte> GetPingMessage()
+        {
+            _pingMessages[1] = _clientConnectionManager.ClientConnections.Count.ToString();
+            _pingMessages[3] = _connectionIds.Count.ToString();
+
+            return ServiceProtocol.GetMessageBytes(
+                new PingMessage
+                {
+                    Messages = _pingMessages
+                });
+        }
+
+        protected override Task OnClientConnectedAsync(OpenConnectionMessage message)
+        {
+            var connection = _clientConnectionFactory.CreateConnection(message, ConfigureContext);
+            connection.ServiceConnection = this;
+
+            if (message.Headers.TryGetValue(Constants.AsrsMigrateFrom, out var from))
+            {
+                connection.Features.Set<IConnectionMigrationFeature>(new ConnectionMigrationFeature(from, ServerId));
+            }
+
+            AddClientConnection(connection, message);
+
+            var scopeProperties = new ClientConnectionScopeProperties() 
+            { 
+                ServiceConnection = this
+                // todo add more properties here, e.g. message.Headers.TryGetValue("Tracing", ...)
+            };
+
+            using (new ServiceConnectionScopeInternal(scopeProperties))
+            {
+                _ = ProcessClientConnectionAsync(connection);

```

Comment:
Does the async local persist in `ProcessClientConnectionAsync` even though it's unset immediately in `Dispose`?

Reply:
It persists only inside `ProcessClientConnectionAsync` and its child execution contexts (e.g. Task.Run, QUWI, etc, including task continuations). We don't want it to leak it outside of `ProcessClientConnectionAsync` because it has a chance to spread to other unrelated execution contexts.

I believe there are some subtle ways it could leak though, for example if we awaited the call to `ProcessClientConnectionAsync` and it happened to complete synchronously, in theory the continuation could run on its child frame and Dispose won't be able to clear out the async local in the original frame... In our case we don't await for the call so `Dispose` is guaranteed to run on the same frame as `using`.

Reply:
To be precise, the async local also persists till the end of `OnClientConnectedAsync` frame but the rest of its execution (after `Dispose`) won't reference scopeProperties so it won't carry large amount of data.

Reply:
I think we should use `Task.Run`:
```c#
_ = Task.Run(async () =>
{
  using (...)
  {
    await ...
  }
});
```

Reply:
I understand the idea is to prevent future changes from accidentally leaking the async local?  Also, we're not awaiting the call to `ProcessClientConnectionAsync` so why should we await it inside `Task.Run`?